src/Rings/MPolyMap/flattenings.jl

```diff
@@ -0,0 +1,247 @@
+    S::AbstractAlgebra.Generic.MPolyRing{RingElemType}
+  ) where {RingElemType <: MPolyQuoLocalizedRingElem}
+  Q = base_ring(S)::MPolyQuoLocalizedRing
+  R = base_ring(Q)::MPolyRing
+  L = localized_ring(Q)::MPolyLocalizedRing
+  A = underlying_quotient(Q)::MPolyQuo
+  U = inverted_set(Q)::AbsMultSet
+  I = modulus(Q)::MPolyLocalizedIdeal
+  kk = coefficient_ring(R)::Field
+
+  # Before building S_flat, we have to create a polynomial 
+  # ring T_flat from which we can pass to the localized quotient.
+  T_flat, _ = PolynomialRing(kk, vcat(symbols(S), symbols(R)))
+  R_to_T_flat = hom(R, T_flat, gens(T_flat)[ngens(S)+1:end])
+
+  T_flat_loc, _ = localization(T_flat, R_to_T_flat(U)) # Will throw if multiplicative set can not be transfered. 
+  L_to_T_flat_loc = hom(L, T_flat_loc, gens(T_flat_loc)[ngens(S)+1:end])
+  
+  I_flat = ideal(T_flat_loc, L_to_T_flat_loc.(gens(I)))
+  S_flat, pr = quo(T_flat_loc, I_flat)
+
+  Q_to_S_flat = hom(Q, S_flat, gens(S_flat)[ngens(S)+1:end])
+  
+  S_to_S_flat = hom(S, S_flat, Q_to_S_flat, gens(S_flat)[1:ngens(S)])
+  S_flat_to_S = hom(S_flat, S, vcat(gens(S), S.(gens(Q))))
+
+  set_attribute!(S, :_flat_to_orig, S_flat_to_S)
+  set_attribute!(S, :_orig_to_flat, S_to_S_flat)
+  set_attribute!(S, :_base_ring_to_flat, Q_to_S_flat)
+
+  return S_flat
+end
+
+@attr function _flatten(
+    S::AbstractAlgebra.Generic.MPolyRing{RingElemType}
+  ) where {RingElemType <: MPolyLocalizedRingElem}
+  L = base_ring(S)::MPolyLocalizedRing
+  R = base_ring(L)::MPolyRing
+  U = inverted_set(L)::AbsMultSet
+  kk = coefficient_ring(R)::Field
+
+  # Before building S_flat, we have to create a polynomial 
+  # ring T_flat from which we can pass to the localized quotient.
+  T_flat, _ = PolynomialRing(kk, vcat(symbols(S), symbols(R)))
+  R_to_T_flat = hom(R, T_flat, gens(T_flat)[ngens(S)+1:end])
+
+  S_flat, _ = localization(T_flat, R_to_T_flat(U)) # Will throw if multiplicative set can not be transfered. 
+  L_to_S_flat = hom(L, S_flat, gens(S_flat)[ngens(S)+1:end])
+  
+  R_to_S_flat = hom(R, S_flat, gens(S_flat)[ngens(S)+1:end])
+  
+  S_to_S_flat = hom(S, S_flat, L_to_S_flat, gens(S_flat)[1:ngens(S)])
+  S_flat_to_S = hom(S_flat, S, vcat(gens(S), S.(gens(R))))
+
+  set_attribute!(S, :_flat_to_orig, S_flat_to_S)
+  set_attribute!(S, :_orig_to_flat, S_to_S_flat)
+  set_attribute!(S, :_base_ring_to_flat, L_to_S_flat)
+
+  return S_flat
+end
+
+### Transfer of multiplicative sets along ring homomorphisms 
+function (phi::MPolyAnyMap{<:MPolyRing, <:MPolyRing, Nothing})(U::MPolyPowersOfElement;
+                                                               check::Bool=true
+                                                              )
+  ambient_ring(U) === domain(phi) || error("multiplicative set does not lay in the domain of the morphism")
+  S = codomain(phi) 
+  SU = MPolyPowersOfElement(S, phi.(denominators(U)))
+  return SU
+end
+
+function (phi::MPolyAnyMap{<:MPolyRing, <:MPolyRing, Nothing})(U::MPolyComplementOfPrimeIdeal;
```

Comment:
Why is that here?

Reply:
This one also gives me a bit of pain, because I don't think there's any such thing in general on the mathematical side. But for some maps and multiplicative sets, one can make sense of what this does. I'm afraid, I have to think about it again to make this more rigorous.

It's here, because it's used in the flattening of polynomial rings with localized rings as coefficient rings.

Reply:
I personally would put it next to where the multiplicative sets are or where the poly ring maps are.